### PR TITLE
Add spectral embedding algorithm

### PR DESCRIPTION
**Pull request by avolkov-intel**
# Description
Add implementation for spectral embedding algorithm in DAAL and oneDAL interfaces for this algorithm

### CODE REVIEW

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(*r, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
+    const size_t binarySearchIterNum = 20;
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:110]: What means a magic number 20?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    // }
+
+    // Find the eigen vectors and eigne values of the matix
+    TArray<algorithmFPType, cpu> eigenvalues(n);
+    DAAL_CHECK_MALLOC(eigenvalues.get());
+
+    services::Status eigen_vectors_st = computeEigenvectorsInplace<algorithmFPType, cpu>(n, x, eigenvalues.get());
+    if (!eigen_vectors_st) {
+        return eigen_vectors_st;
+    }
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:174]:
```suggestion
    DAAL_CHECK_STATUS(status, computeEigenvectorsInplace<algorithmFPType, cpu>(n, x, eigenvalues.get()));
```

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_kernel.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+{
+
+using namespace daal::data_management;
+using namespace daal::services;
+
+
+struct KernelParameter : daal::algorithms::Parameter
+{
+    size_t numEmb = 1;
+    size_t numNeighbors = 1;
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_kernel.h:52]: In DAAL we try to use full names where possible.
```suggestion
    size_t numberOfEmbeddings = 1;
    size_t numberOfNeighbors = 1;
```

File: cpp/oneapi/dal/algo/spectral_embedding/backend/cpu/compute_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        
+        daal::services::SharedPtr<NumericTable> daal_input, daal_output;
+        array<Float> arr_output;
+        arr_output = array<Float>::empty(n * k);
+        daal_output = interop::convert_to_daal_homogen_table(arr_output, n, k);
+        parameter_t daal_param;
+
+        daal_param.numEmb = k;
+        if (desc.get_num_neighbors() < 0) {
+            daal_param.numNeighbors = n - 1;
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/spectral_embedding/backend/cpu/compute_kernel.cpp:71]: Why `n-1` ? and what if number of neighbors == 0?

File: cpp/oneapi/dal/algo/spectral_embedding/common.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace v1 {
+
+template <typename Task>
+class descriptor_impl : public base {
+public:
+    explicit descriptor_impl() {}
+
+    std::int64_t embedding_dim = 0;
+    std::int64_t num_neighbors = -1;
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/spectral_embedding/common.cpp:44]: Let's align with scikit and use `components` for the dimension of projected space. Also, in oneDAL we usually use `count`, not `dim` or `num` in naming.
In DAAL it would be `numberOfComponents`, etc.
```suggestion
    std::int64_t component_count = 0;
    std::int64_t neighbor_count = -1;
```

File: cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        INFO("run compute");
+        auto compute_result = this->compute(desc, data_);
+        check_compute_result(compute_result);
+        std::cout << "Output" << std::endl;
+        // std::cout << compute_result.get_embedding() << std::endl;
+    }
+
+    void check_compute_result(const spectral_embedding::compute_result<>& result) {
+        array<Float> data_arr = row_accessor<const Float>(data_).pull({ 0, -1 });
+    }
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp:76]: Please add comparison with 'golden' data computed with sklearn.

File: cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+
+    void test_default() {
+        std::cout << "Input" << std::endl;
+        // std::cout << data_ << std::endl;
+        auto desc =
+            get_descriptor(sp_emb::result_options::embedding);
+        //desc.set_embedding_dim(5);
+        //desc.set_num_neighbors(4);
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp:66]: Let's compute the number of neighbors by the formula provided in the original request.

**Review by Vika-F — COMMENTED** [review]
Not bad for the initial version.
Please decompose the compute function into smaller kernels. I'd move binary search part and Laplassian part into separate functions.
My other comments are below.

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i (hunk 1 above)
Comment (avolkov-intel) [reply cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:110]: For now it is actually a magic number but it should be log(1/eps) where eps is floating point number precision

File: cpp/oneapi/dal/algo/spectral_embedding/backend/cpu/compute_kernel.cpp (hunk 4 above)
Comment (avolkov-intel) [reply cpp/oneapi/dal/algo/spectral_embedding/backend/cpu/compute_kernel.cpp:71]: For example, if number of neighbors is specified to -1 it means that no pruning is actually needed, so all the edges should be kept

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i (hunk 1 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:110]: It would be better then to define pre-computed constants for `float` and `double`, add a description, and use those constants in the computations.
For example, like it is done here: https://github.com/oneapi-src/oneDAL/blob/main/cpp/daal/src/algorithms/em/em_gmm_dense_default_batch_impl.i#L144
lines 144-157

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    {
+        L    = 0; // min possible cos distance
+        R    = 2; // max possible cos distance
+        lcnt = 0; // number of elements with cos distance <= L
+        rcnt = n; // number of elements with cos distance <= R
+        for (size_t ij = 0; ij < binarySearchIterNum; ++ij)
+        {
+            M   = (L + R) / 2;
+            cnt = 0;
+            // Calculate the number of elements in the row with value <= M
```
Comment (icfaust) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:112]: Just from a high level perspective, it looks like this code will run O(binarySearchIterNum*n) to try and find the proper M which yields the right number of neighbors. binarySearchIterNum will be log(1/eps) (from here https://github.com/oneapi-src/oneDAL/pull/2875/files#r1720389061) Would a binary search tree be faster?  At some point also wouldn't just sorting the array may also be faster (when log(n) < log(1/eps))?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            {
+                x[i * n + j] = 0.0;
+            }
+        }
+        // fill the diagonal of matrix with zeros
+        x[i * n + i] = 0;
+    }
+
+    // Create Laplassian matrix
+    for (size_t i = 0; i < n; ++i)
```
Comment (icfaust) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:153]: Dumb question, I sort assumed the nearest cosine neighbors from above created almost an adjacency matrix which could be used here to create the laplacian via L=D-A.  I guess I am trying to understand the creation math here, could you send me a link what definition you used here?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    //     }
+    //     std::cout << std::endl;
+    // }
+
+    // Fill the output matrix with eigen vectors corresponding to the smallest eigen values
+    WriteOnlyRows<algorithmFPType, cpu> embedMatrix(embeddingTable, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(embedMatrix);
+    algorithmFPType * embed = embedMatrix.get();
+
+    for (int i = 0; i < k; ++i)
```
Comment (icfaust) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:197]: Maybe also a todo for a parallel for here?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        DAAL_CHECK_STATUS(status, cosDistanceKernel.compute(0, &a0, 0, &covOutput, nullptr));
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(covOutput, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
```
Comment (icfaust) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:99]:
```suggestion
    // Use binary search to find such d that the number of vertices having distance <= d is filtNum
```

**Review by icfaust — COMMENTED** [review]
Out of the blue questions about the algo implementation.  Nothing to be forced, I just want to start a conversation with you and the other reviewers.

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    //TArray<algorithmFPType, cpu> eigenvalues(n);
+    //DAAL_CHECK_MALLOC(eigenvalues.get());
+    WriteRows<algorithmFPType, cpu> eigenValuesBlock(eigenvalues, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(eigenValuesBlock);
+    algorithmFPType * eigenValuesPtr = eigenValuesBlock.get();
+
+    status |= computeEigenvectorsInplace<algorithmFPType, cpu>(n, x, eigenValuesPtr);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    // std::cout << "Eigen vectors: " << std::endl;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:184]: Just a reminder to delete unused code

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    // Create Laplassian matrix
+    for (size_t i = 0; i < n; ++i)
+    {
+        for (size_t j = 0; j < i; ++j)
+        {
+            algorithmFPType val = (x[i * n + j] + x[j * n + i]) / 2;
+            x[i * n + j]        = -val;
+            x[j * n + i]        = -val;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:159]: I'm a little bitconfused with inconsistency - in this case you are assigning values to the elements of matrix, but for the other part you are incrementing the symmetric ones. Does this matrix represent two triangular ones or should operations be actually symmetric?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i (hunk 10 above)
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:197]: Should it be `size_t` instead?

File: cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        data_ = data_df.get_table(this->get_policy(), this->get_homogen_table_id());
+    }
+
+    void test_gold_input(Float tol = 1e-5) {
+        constexpr std::int64_t n = 8;
+        constexpr std::int64_t p = 4;
+        constexpr std::int64_t neighbor_count = 5;
+        constexpr std::int64_t component_count = 4;
+
+        constexpr Float data[n * p] = { 0.49671415,  -0.1382643,  0.64768854,  1.52302986,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/spectral_embedding/test/fixture.hpp:67]: It seems this data is auto-generated. Please consider using rng to generate it.

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i (hunk 13 above)
Comment (icfaust) [reply cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:159]: I am/was too. Hopefully a bit of clarity to help in a code comment.

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+            {
+                break;
+            }
+        }
+        // create edges for the closest neighbors
+        for (size_t j = 0; j < n; ++j)
+        {
+            if (x[i * n + j] <= R)
+            {
+                x[i * n + j] = 1.0;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:141]:
```suggestion
                x[i * n + j] = algorithmFPType(1);
```

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            }
+            else
+            {
+                x[i * n + j] = 0.0;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:145]:
```suggestion
                x[i * n + j] = algorithmFPType(0);
```

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+    // Compute cosine distances matrix
+    {
+        auto cosDistanceKernel = cosine_distance::internal::DistanceKernel<algorithmFPType, cosine_distance::Method::defaultDense, cpu>();
+        DAAL_CHECK_STATUS(status, cosDistanceKernel.compute(0, &a0, 0, &covOutput, nullptr));
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(covOutput, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:96]: I believe that according to SDL somewhere around here should be several `DAAL_ASSERT_...` statements to check if your access to the memory is contained.

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:98]: I would recommend to initialize these variables. It's not a C code =D

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
+    const size_t binarySearchIterNum = 20;
+    // TODO: add parallel_for
+    for (size_t i = 0; i < n; ++i)
+    {
+        L    = 0; // min possible cos distance
+        R    = 2; // max possible cos distance
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i:105]: As far as I remember at least some of these variables are actually `float`s. Should we initialize them accordingly?

File: cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_kernel.h [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace internal
+{
+
+using namespace daal::data_management;
+using namespace daal::services;
+
+struct KernelParameter : daal::algorithms::Parameter
+{
+    size_t numberOfEmbeddings = 1;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_kernel.h:51]: Should it be `DAAL_INT`? Personally I have nothing against using `size_t` just caring for the consistency.

File: examples/oneapi/cpp/source/spectral_clustering/spectral_clustering_pipeline.cpp [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+* limitations under the License.
+*******************************************************************************/
+
+#include "example_util/utils.hpp"
+#include "oneapi/dal/algo/kmeans.hpp"
+#include "oneapi/dal/algo/kmeans_init.hpp"
+#include "oneapi/dal/algo/spectral_embedding.hpp"
+#include "oneapi/dal/io/csv.hpp"
+#include <algorithm>
+#include <math.h>
```
Comment (KulikovNikita) [inline examples/oneapi/cpp/source/spectral_clustering/spectral_clustering_pipeline.cpp:23]:
```suggestion
#include <cmath>
```

Please use a [regular C++ headers](https://en.cppreference.com/w/cpp/header/cmath)
